landlab/utils/source_tracking_algorithm.py

```diff
@@ -0,0 +1,234 @@
+import numpy as np
+import copy
+from collections import Counter
+
+# %%
+def convert_arc_flow_directions_to_landlab_node_ids(grid, flow_dir_arc):
+    """Convert Arc flow_directions to RasterModelGrid node ids
+
+    This function receives flow directions (D8) from ESRI ArcGIS and converts
+    them to Landlab's RasterModelGrid node id. ESRI ArcGIS D8 flow directions
+    are either of the eight valid output directions relating to the eight
+    adjacent cells into which flow could travel. The valid output directions
+    are powers of 2 starting from 2^0 (1) in the Eastern neighbor going
+    clockwise to 2^7 (128) at Northeastern neighbor. For more information
+    refer 'http://pro.arcgis.com/en/pro-app/tool-reference/spatial-analyst/
+    how-flow-direction-works.htm'
+    
+    Parameters:
+    ----------
+    grid: RasterModelGrid
+        A grid.
+    flow_dir_arc: ndarray of int, shape (n_nodes, )
+        flow directions derived from ESRII ArcGIS.
+    
+    Returns:
+    -------
+    receiver_nodes: ndarray of int, shape (n_nodes, )
+        downstream node at each node. Note that this array gives the
+        receiver nodes only for the core nodes. For non-core
+        nodes, a zero is used.  
+    """
+    r_arc_raw = np.log2(flow_dir_arc)
+    r_arc_raw = r_arc_raw.astype('int')
+    neigh_ = grid.neighbors_at_node
+    diag_ = grid.diagonals_at_node
+    neigh_ = np.fliplr(neigh_)
+    diag_ = np.fliplr(diag_)
+    a_n = np.hsplit(neigh_, 4)
+    a_d = np.hsplit(diag_, 4)
+    neighbors = np.hstack((a_n[-1], a_d[0], a_n[0], a_d[1], a_n[1], a_d[2],
+                           a_n[2], a_d[3]))
+    # Now neighbors has node ids of neighboring nodes in cw order starting at
+    # right, hence the order of neighbors = [r, br, b, bl, l, tl, t, tr]
+    receiver_nodes = np.zeros(grid.number_of_nodes, dtype=int)
+    receiver_nodes[grid.core_nodes] = np.choose(r_arc_raw[grid.core_nodes],
+                                   np.transpose(neighbors[grid.core_nodes]))
+    return (receiver_nodes)
+
+
+# %%
+# Source Routing Algorithm
+# Note 1: This algorithm works on core nodes only because core nodes
+# have neighbors that are real values and not -1s.
+# Note 2: Nodes in the following comments in this section refer to core nodes.
+def track_source(grid, hsd_ids, flow_directions=None):
+    """Track all contributing upstream core nodes for each core node
+
+    This algorithm traverses the grid, and records all upstream nodes for each
+    node, given the flow directions. Alternatively, flow directions can be
+    attached to the grid as a field. This function works similar to
+    flow accumulation routine, that include even a single contributing
+    upstream node, but is executed at each node in the Modeling Domain (MD).
+    The algorithm uses brute force method and is time intensive. This
+    function might not work on huge watersheds (we haven't derived the
+    maximum size of the watershed that this function can execute).
+
```

Comment:
This description is better, but it is still a bit confusing. I've taken a stab at a revision of the first paragraph here (I turned it into two paragraphs):

This algorithm traverses the grid based on information of flow directions at nodes and at every node identifies _all_ the nodes upstream of a given node. The algorithm creates a dictionary with an entry for each node; a node's entry in the dictionary will contain a list with the node_ids of all upstream nodes. Thus this method permits identification of the source area contributing to each and every node in the model grid. This function is different from a standard flow accumulation routine in that it not only calculates the amount of flow at each node, but records the IDs of _all_ upstream nodes. However, similar to a standard flow accumulation routine, it produces an at_node array of the amount of flow passing through the node. It also differs from a standard flow accumulation routing in that it permits the mapping of flow inputs from a coarser grid to to a finer model grid. 

In its present implementation the algorithm has not been optimized for efficient time use. Its methods are brute force and it should be expected to be time intensive. It is not recommended to be run frequently in a modeling exercise. Due to its intensive nature, this algorithm may fail with large watersheds (a present, the development team has not derived a maximum stable watershed size).